poh/src/poh_recorder.rs

```diff
@@ -539,6 +573,58 @@ impl PohRecorder {
         Ok(())
     }
 
+    fn report_poh_timing_point_by_tick(&self) {
+        // reaching the end of the slot
+        if self.tick_height % self.ticks_per_slot == 0 {
+            let slot = self.slot_for_tick_height(self.tick_height);
+            if let Some(ref sender) = self.poh_timing_point_sender {
+                trace!("PohTimingPoint:End {}", slot);
+                let _ = sender.try_send((
+                    slot,
+                    None,
+                    PohTimingPoint::PohSlotEnd(solana_sdk::timing::timestamp()),
+                ));
+            }
+        }
+
+        // beginning of a slot
+        if self.tick_height % self.ticks_per_slot == 1 {
+            let slot = self.slot_for_tick_height(self.tick_height);
+            if let Some(ref sender) = self.poh_timing_point_sender {
+                trace!("PohTimingPoint:Start {}", slot);
+                let _ = sender.try_send((
+                    slot,
+                    None,
+                    PohTimingPoint::PohSlotStart(solana_sdk::timing::timestamp()),
+                ));
+            }
+        }
+    }
+
+    fn report_poh_timing_point_by_working_bank(&self) {
+        if let Some(slot) = self.working_slot() {
+            if let Some(ref sender) = self.poh_timing_point_sender {
+                trace!("PohTimingPoint:End {}", slot);
+                let _ = sender.try_send((
+                    slot,
+                    None,
+                    PohTimingPoint::PohSlotEnd(solana_sdk::timing::timestamp()),
+                ));
+            }
+        }
+    }
+
+    fn report_poh_timing_point(&self) {
+        // send poh slot end timing point
+        if self.working_bank_end() && self.has_bank() {
```

Comment:
`self.has_bank()` seems redundant here? `working_bank_end()` already checks the working_bank is not None